### DISCUSSION TITLE
Stdin pipe not closed in `Msg.WriteToSendmailWithContext`

### DISCUSSION THREAD

**Discussion by inliquid** [wneessen/go-mail#5] [General]
Not sure if it's an issue or my bad understanding of how cli MUA works, so I open a question here in discussions. In my target setup I'm using `exim` to send mails to users. I don't have working `exim` setup on my dev machine, so I decided to write a small program to log all calls to "exim" to a file while testing everything. 

Here is how it looks like:
```go
func main() {
	logger.Info.Println("-> Program started")
	defer logger.Info.Println("<- Program exited")

	logger.Info.Println("Args:", os.Args)

	s := bufio.NewScanner(os.Stdin)
	for s.Scan() {
		if s.Text() == "." {
			logger.Info.Println(`Got "." ending, stopped reding`)
			break
		}
		logger.Info.Println("Read line:", s.Text())
	}
	if err := s.Err(); err != nil {
		logger.Error.Panicln("Can't read os.Stdin:", err)
	}
}
```
It uses own logger based on standard logger. This program is compiled as `exim` and put in `PATH`. Not sure if original `exim` interprets "." as message ending, but added this  just in case.

What I observed, effectively, is that `WriteToSendmailWithContext` hangs unless there is timeout set on given context. In last case such call return an error. In logs I can see that reading from `os.Stdin` stops somewhere before last line and that's it:
```
[INFO] 2022/05/24 18:11:27.876387 -> Program started
[INFO] 2022/05/24 18:11:27.876387 Args: [exim -oi -t]
[INFO] 2022/05/24 18:11:27.876387 Read line: Date: Tue, 24 May 2022 18:11:27 +0300
[INFO] 2022/05/24 18:11:27.876387 Read line: MIME-Version: 1.0
[INFO] 2022/05/24 18:11:27.876387 Read line: Message-ID: <20156.7468985450408456538.1653405087@inflow-pc>
[INFO] 2022/05/24 18:11:27.876387 Read line: Subject: Test message
[INFO] 2022/05/24 18:11:27.876387 Read line: User-Agent: go-mail v0.1.7 // https://github.com/wneessen/go-mail
[INFO] 2022/05/24 18:11:27.876387 Read line: X-Mailer: go-mail v0.1.7 // https://github.com/wneessen/go-mail
[INFO] 2022/05/24 18:11:27.876387 Read line: From: ***
[INFO] 2022/05/24 18:11:27.876387 Read line:  ***
[INFO] 2022/05/24 18:11:27.876387 Read line: To: ***
[INFO] 2022/05/24 18:11:27.876387 Read line: Content-Type: text/html; charset=UTF-8
[INFO] 2022/05/24 18:11:27.876387 Read line: Content-Transfer-Encoding: quoted-printable
[INFO] 2022/05/24 18:11:27.876387 Read line: 
```

I looked at implementation and I think problem (if I get it right) is that there is no indication that message is done. There is neither a single dot on a line "." (not sure if it's a correct way of terminating message on `sendmail`-like cli) , nor closed stdin, because its `Close` method is called in deferred closure. Instead, I think `si.Close()` should be called right before `ec.Wait()` so that stdin would be closed just after message written to it, but before program is exited.

**Comment by wneessen** [comment]
Thanks for the report. You are right. While the sendmail binary will not be waiting for a "." we should force the closing of STDIN so that the connection to receiving end can be completed with a EOF.

I've added the forcefully close in commit 070eb39132e0bc155f5be5fc593dc5560f66cd96 and merged it into main. Please give it a try. I've also realized that sendmail is located in /usr/sbin on most common unix systems therefore the default path has been adjusted accordingly.

**Comment by inliquid** [comment]
Thank you for quick update! May I ask to tag latest changes so that go modules system will see updates without the need to explicitly pull it via `go get <...>@master`

**Reply by wneessen** [reply]
Done: https://github.com/wneessen/go-mail/releases/tag/v0.1.9-beta.1

**Reply by inliquid** [reply]
It should be in semver format like `v0.1.9`, it won't pick it with suffix unfortunately.

**Reply by wneessen** [reply]
Corrected: https://github.com/wneessen/go-mail/releases/tag/v0.1.9
